awx/main/models/notifications.py

```diff
@@ -200,56 +246,115 @@ def get_absolute_url(self, request=None):
+                            {'playbook_counts': ['play_count', 'task_count']},
+                            {'summary_fields': [{'inventory': ['id', 'name', 'description', 'has_active_failures',
+                                                               'total_hosts', 'hosts_with_active_failures', 'total_groups',
+                                                               'groups_with_active_failures', 'has_inventory_sources',
+                                                               'total_inventory_sources', 'inventory_sources_with_failures',
+                                                               'organization_id', 'kind']},
+                                                {'project': ['id', 'name', 'description', 'status', 'scm_type']},
+                                                {'project_update': ['id', 'name', 'description', 'status', 'failed']},
+                                                {'job_template': ['id', 'name', 'description']},
+                                                {'unified_job_template': ['id', 'name', 'description', 'unified_job_type']},
+                                                {'instance_group': ['name', 'id']},
+                                                {'created_by': ['id', 'username', 'first_name', 'last_name']},
+                                                {'labels': ['count', 'results']}]}]
+
+    @classmethod
+    def context_stub(cls):
+        """Returns a stub context that has the same structure as the context that will
+        actually be used to render a notification message (useful for validation)"""
+        context = {'job': {},
+                   'job_friendly_name': 'stub',
+                   'url': 'stub'}
+
+        def build_context_for_fields(node, fields):
+            for f in fields:
+                if type(f) is dict:
+                    subname, subfields = f.copy().popitem()
+                    node[subname] = {}
+                    build_context_for_fields(node[subname], subfields)
+                else:
+                    node[f] = 'stub'
+        build_context_for_fields(context['job'], cls.JOB_FIELDS_WHITELIST)
+
+        return context
+
     def get_notification_templates(self):
         raise RuntimeError("Define me")
 
     def get_notification_friendly_name(self):
         raise RuntimeError("Define me")
 
-    def _build_notification_message(self, status_str):
+    def notification_data(self):
+        raise RuntimeError("Define me")
+
+    def build_notification_message(self, nt, status):
+        env = sandbox.ImmutableSandboxedEnvironment()  # TODO: sub-class to define unsafe attrs
+
+        from awx.api.serializers import UnifiedJobSerializer  # Need to import here (To avoid circular import?)
+        job_serialization = UnifiedJobSerializer(self).to_representation(self) # TODO: Filter job serialization to only include whitelisted fields
+        context = dict(job=job_serialization,
```

Comment:
It would be good if the API validation + the building of the messages share the _same_ code

Reply:
yep, I've only done half of the work (validation), need to do the building of the message (the other half) next. That's in progress.